app/javascript/packages/captcha-submit-button/captcha-submit-button-element.ts

```diff
@@ -0,0 +1,70 @@
+class CaptchaSubmitButtonElement extends HTMLElement {
+  connectedCallback() {
+    this.button.addEventListener('click', (event) => this.handleButtonClick(event));
+  }
+
+  get isExempt(): boolean {
+    return this.hasAttribute('exempt');
+  }
+
+  get button(): HTMLButtonElement {
+    return this.querySelector('button')!;
+  }
+
+  get tokenInput(): HTMLInputElement {
+    return this.querySelector('[type=hidden]')!;
+  }
+
+  get form(): HTMLFormElement | null {
+    return this.closest('form');
+  }
+
+  get recaptchaSiteKey(): string | null {
+    return this.getAttribute('recaptcha-site-key');
+  }
+
+  get recaptchaAction(): string {
+    return this.getAttribute('recaptcha-action')!;
+  }
+
+  submit() {
+    this.form?.submit();
+  }
+
+  invokeChallenge() {
+    grecaptcha.ready(async () => {
```

Comment:
this doesn't delay initialization until this point, right? it just means that in case we challenge too early, we'll wait for initialization?

Reply:
> this doesn't delay initialization until this point, right? it just means that in case we challenge too early, we'll wait for initialization?

I guess it depends what's meant by initialization. The main script will always be present on the page since the page load, which may have its own initialization behavior. I'm not entirely sure what's involved in this `ready` behavior, though the fact that it's asynchronous would lead me to believe that it could involve some delay. But in practice it doesn't appear to be noticeable, and this implementation follows [similar examples in the documentation for invoking the challenge in response to a button click](https://developers.google.com/recaptcha/docs/v3#programmatically_invoke_the_challenge).